### PR TITLE
Add functional API

### PR DESCRIPTION
**Pull request by encukou**
This adds functional API, proposed for CPython here: https://github.com/python/cpython/issues/116608

I tested locally using `tox`, which probably doesn't match the CI checks.

### CODE REVIEW

**Comment by encukou** [comment]
“Collateral” tests fail with:
```
importlib_resources/functional.py (62.5%): Missing lines 3,5,8,11,16,23,28,35,40,48,63,78
importlib_resources/tests/test_functional.py (98.2%): Missing lines 7,9
```

These are top-level statements (mostly `def`), and an alternate import for old Python versions. I'm not familiar with the coverage tool, and would appreciate a hint on how to fix this. (Or feel free to push a fix directly to the PR.)

File: importlib_resources/functional.py
Comment (layday) [inline importlib_resources/functional.py:1]: Should this module be private or should importing from `importlib_resources.functional` be allowed?

Comment (layday) [reply importlib_resources/functional.py:1]: (I don't recall if it was public before.)

Comment (jaraco) [reply importlib_resources/functional.py:1]: Good call. Private is probably better.

**Pull request merged**